compiler/rustc_parse/src/parser/expr.rs

```diff
@@ -857,30 +859,52 @@ impl<'a> Parser<'a> {
             components.push(IdentLike(ident_like));
         }
 
-        // FIXME: Make the span more precise.
+        // With proc macros the span can refer to anything. Validate that it refers to something
+        // identical to our float literal, only then we can safely break it into parts.
         let span = self.token.span;
+        let span_is_good = || self.span_to_snippet(span).as_deref() == Ok(float_str).as_deref();
+
         match &*components {
             // 1e2
             [IdentLike(i)] => {
                 self.parse_tuple_field_access_expr(lo, base, Symbol::intern(&i), suffix, None)
             }
             // 1.
             [IdentLike(i), Punct('.')] => {
+                let (ident_span, dot_span) = if span_is_good() {
```

Comment:
Can we rename `span_is_good` to `span_is_not_through_macro` or something else more descriptive? Even though the diff is easy to follow, it'll be hard to understand what this means on isolation, and the comment explaining why is not next to the closure.

Reply:
I've expanded the comment a bit.
(I don't think it's not about macros exactly, just about dangers of arithmetic with arbitrary spans.)